src/custom_instance_manager.ts

```diff
@@ -0,0 +1,132 @@
+    for (let i = 0; i < length; i++) {
+        result += characters.charAt(Math.floor(Math.random() * charactersLength));
+    }
+    return result;
+}
+
+export interface CustomInstanceManagerOptions {
+    isDryRun: boolean;
+}
+
+export interface CustomCloudInstance {
+    instanceId: string;
+    displayName: string;
+    cloudStatus: string;
+}
+
+export default class CustomInstanceManager {
+    private isDryRun: boolean;
+    private instances: Record<string, CustomCloudInstance>;
+    constructor(options: CustomInstanceManagerOptions) {
+        this.isDryRun = options.isDryRun;
+
+        this.launchInstances = this.launchInstances.bind(this);
+    }
+
+    async launchInstances(
+        ctx: Context,
+        group: InstanceGroup,
+        groupCurrentCount: number,
+        quantity: number,
+    ): Promise<Array<string | boolean>> {
+        ctx.logger.info(`[custom] Launching a batch of ${quantity} instances in group ${group.name}`);
+
+        const indexes: Array<number> = [];
+        for (let i = 0; i < quantity; i++) {
+            indexes.push(i);
+        }
+
+        const result = await Promise.all(
+            indexes.map(async (index) => {
+                return this.launchInstance(ctx, index, group);
+            }),
+        );
+        ctx.logger.info(`Finished launching all the instances in group ${group.name}`);
+
+        return result;
+    }
+
+    async launchInstance(ctx: Context, index: number, group: InstanceGroup): Promise<string | boolean> {
+        const groupName = group.name;
+        const groupInstanceConfigurationId = group.instanceConfigurationId;
+
+        const displayName = groupName + '-' + makeRandomString(5);
+
+        ctx.logger.info(`[custom] Launching instance number ${index + 1} in group ${groupName} with properties`, {
+            groupName,
+            displayName,
+            groupInstanceConfigurationId,
+        });
+
+        if (this.isDryRun) {
+            ctx.logger.info(`[custom] Dry run enabled, skipping the instance number ${index + 1} launch`);
+            return true;
+        }
+        try {
+            const launchResponse = await this.execLaunch({
+                displayName,
+                groupName,
+                region: group.region,
+                type: group.type,
+            });
+            ctx.logger.info(
+                `[custom] Got launch response for instance number ${
+                    index + 1
+                } in group ${groupName}: ${launchResponse}`,
+            );
+
+            return launchResponse;
+        } catch (err) {
+            ctx.logger.error(
+                `[custom] Failed launching instance number ${index + 1} in group ${groupName} with err ${err}`,
+                { err },
+            );
+            return false;
+        }
+    }
+
+    async getInstances(): Promise<CustomCloudInstance[]> {
+        // TODO we might need another script to fetch the current instance status
```

Comment:
The purpose of this method is to read from the cloud provider the instances that are actually in the cloud (not in the autoscaler).
We are using this info for diagnostics (see the 'report' endpoints) - we view what the cloud sees vs what the autoscaler knows and compare the 2 views.
We monitor the differences that we see (for example maybe there are instances only in the cloud, but which are invisible to the autoscaler) and this allows anyone to act when such difference occurs.

I think this method should call a script to retrieve that list of cloud instances. In case you are not planning to implement this now, you can just return [] as you did initially and add a TODO for later implementation.